src/ucan-cloud/ucan-gateway.ts

```diff
@@ -0,0 +1,367 @@
+import { exception2Result, KeyedResolvOnce, Result, URI } from "@adviser/cement";
+import { bs, getStore, Logger, SuperThis, ensureSuperLog, NotFoundError, ensureLogger, rt } from "@fireproof/core";
+import { DID } from "@ucanto/core";
+import { ConnectionView, Principal } from "@ucanto/interface";
+import { Absentee } from "@ucanto/principal";
+import * as DidMailto from "@web3-storage/did-mailto";
+import * as W3 from "@web3-storage/w3up-client";
+import { Service as W3Service } from "@web3-storage/w3up-client/types";
+import { AgentData, AgentDataExport } from "@web3-storage/access/agent";
+
+import { CID } from "multiformats";
+
+import * as Client from "./client";
+import { Service } from "./types";
+import { exportDelegation } from "./common";
+import stateStore from "./store/state";
+
+export class UCANGateway implements bs.Gateway {
+  readonly sthis: SuperThis;
+  readonly logger: Logger;
+
+  inst?: {
+    clock: Client.Clock;
+    email: `${string}@${string}`;
+    server: Principal;
+    service: ConnectionView<Service>;
+    w3: W3.Client;
+  };
+
+  constructor(sthis: SuperThis) {
+    this.sthis = ensureSuperLog(sthis, "UCANGateway");
+    this.logger = this.sthis.logger;
+  }
+
+  async buildUrl(baseUrl: URI, key: string): Promise<Result<URI>> {
+    return Result.Ok(baseUrl.build().setParam("key", key).URI());
+  }
+
+  async start(baseUrl: URI): Promise<Result<URI>> {
+    const result = await exception2Result(() => this.#start(baseUrl));
+    if (result.isErr()) this.logger.Error().Msg(result.Err().message);
+    return result;
+  }
+
+  async #start(baseUrl: URI): Promise<URI> {
+    const dbName = baseUrl.getParam("name");
```

Comment:
it("BuildURI getParamsResult fail", () => {
    const url = BuildURI.from("http://host/bla/blub?name=test&store=meta&key=%40bla");
    const rParams = url.getParamsResult("key", "name", "store", "key2", "name2");
    expect(rParams.Err().message).toEqual(`missing parameters: key2,name2`);
    const rParams2 = url.getParamsResult(
      "key",
      "name",
      "store",
      "key2",
      "name2",
      (...keys: string[]) => `keys not found: ${keys.join(",")}`,
    );
    expect(rParams2.Err().message).toEqual(`keys not found: key2,name2`);
  });